ouroboros-network/src/Ouroboros/Network/Protocol/TxSubmission/Examples.hs

```diff
@@ -0,0 +1,377 @@
+                                            | tx <- unackedExtra ]
+              remainingTxs'  = drop (fromIntegral reqNo) remainingTxs
+              txIdAndSize tx = (txId tx, txSize tx)
+
+          return $! case (blocking, unackedExtra) of
+            (TokBlocking, []) ->
+              SendMsgDone ()
+
+            (TokBlocking, tx:txs) ->
+              SendMsgReplyTxIds
+                (BlockingReply (fmap txIdAndSize (tx :| txs)))
+                (client unackedSeq'' unackedMap'' remainingTxs')
+
+            (TokNonBlocking, txs) ->
+              SendMsgReplyTxIds
+                (NonBlockingReply (map txIdAndSize txs))
+                (client unackedSeq'' unackedMap'' remainingTxs')
+
+        recvMsgRequestTxs :: [txid]
+                          -> m (ClientStTxs txid tx m ())
+        recvMsgRequestTxs txids = do
+          traceWith tracer (EventRecvMsgRequestTxs unackedSeq unackedMap
+                                                   remainingTxs txids)
+          case [ txid | txid <- txids, txid `Map.notMember` unackedMap ] of
+            [] -> pure (SendMsgReplyTxs txs client')
+              where
+                txs         = map (unackedMap Map.!) txids
+                client'     = client unackedSeq unackedMap' remainingTxs
+                unackedMap' = foldr Map.delete unackedMap txids
+
+            missing -> fail $ "txSubmissionClientConst.recvMsgRequestTxs: "
+                           ++ "requested missing TxIds: " ++ show missing
+
+
+--
+-- Example server
+--
+
+data TraceEventServer txid tx =
+     EventRequestTxIdsBlocking  (ServerState txid tx) Word16 Word16
+   | EventRequestTxIdsPipelined (ServerState txid tx) Word16 Word16
+   | EventRequestTxsPipelined   (ServerState txid tx) [txid]
+
+deriving instance (Show txid, Show tx) => Show (TraceEventServer txid tx)
+
+data ServerState txid tx = ServerState {
+       -- | The number of transaction identifiers that we have requested but
+       -- which have not yet been replied to. We need to track this it keep
+       -- our requests within the limit on the number of unacknowledged txids.
+       --
+       requestedTxIdsInFlight :: Word16,
+
+       -- | Those transactions (by their identifier) that the client has told
+       -- us about, and which we have not yet acknowledged. This is kept in
+       -- the order in which the client gave them to us. This is the same order
+       -- in which we submit them to the mempool (or for this example, the final
+       -- result order). It is also the order we acknowledge in.
+       unacknowledgedTxIds :: Seq txid,
+
+       -- | Those transactions (by their identifier) that we can request. These
+       -- are a subset of the 'unacknowledgedTxIds' that we have not yet
+       -- requested. This is not ordered to illustrate the fact that we can
+       -- request txs out of order. We also remember 
+       availableTxids      :: Map txid TxSizeInBytes,
+
+--       txBytesInFlight     :: TxSizeInBytes,
```

Comment:
^ It will make sense to track the bytes in flight, but indeed this can be left for later. Maybe drop a todo.

Reply:
I've removed it. This will be included in the real version.